src/iguana/algorithms/clas12/LeptonIDFilter/Algorithm.cc

```diff
@@ -0,0 +1,172 @@
+        auto lepton_pindex = FindLepton(particleBank);
+        auto lepton_vars=GetLeptonIDVariables(lepton_pindex,particleBank,calorimeterBank);
+        lepton_vars.score=CalculateScore(lepton_vars);
+
+        return Filter(lepton_vars.score);
+        });
+
+    // dump the modified bank
+    ShowBank(particleBank, Logger::Header("OUTPUT PARTICLES"));
+    
+  }
+
+
+  int LeptonIDFilter::FindLepton(hipo::bank const& particle_bank) const{  
+    int lepton_pindex= -1;
+    for(int row = 0; row < particle_bank.getRows(); row++) {
+        auto status = particle_bank.getShort("status", row);
+        if(particle_bank.getInt("pid", row) == o_pid && abs(status)>=2000 && abs(status)<4000) {
+          lepton_pindex=row;
+          break;
+        }
+      }
+      if(lepton_pindex >= 0)
+        m_log->Debug("Found lepton: pindex={}", lepton_pindex);
+      else
+        m_log->Debug("Lepton not found");
+      return lepton_pindex;
+  }
+
+  LeptonIDVars LeptonIDFilter::GetLeptonIDVariables(int const plepton, hipo::bank const& particle_bank, hipo::bank const& calorimeter_bank) const{
+
+      double px = particle_bank.getFloat("px", plepton);
+      double py = particle_bank.getFloat("py", plepton);
+      double pz = particle_bank.getFloat("pz", plepton);
+      double E = std::sqrt(std::pow(px, 2) + std::pow(py, 2) + std::pow(pz, 2) + std::pow(0.000511, 2));
+      ROOT::Math::PxPyPzMVector vec_lepton(px, py, pz, E);
+
+      LeptonIDVars lepton;
+
+      lepton.P    =vec_lepton.P();
+      lepton.Theta=vec_lepton.Theta();
+      lepton.Phi  =vec_lepton.Phi();
+
+      m_log->Debug("Variables obtained from particle bank");
+
+
+      lepton.m2pcal=-1;
+      lepton.m2ecin=-1;
+      lepton.m2ecout=-1;
+
+      for(int row = 0; row < calorimeter_bank.getRows(); row++) {
+          auto pindex = calorimeter_bank.getShort("pindex",row);
+          auto layer  = calorimeter_bank.getByte("layer",row);
+          auto energy = calorimeter_bank.getFloat("energy",row);
+          auto m2u    = calorimeter_bank.getFloat("m2u",row);
+          auto m2v    = calorimeter_bank.getFloat("m2v",row);
+          auto m2w    = calorimeter_bank.getFloat("m2w",row);
+
+          if(pindex==plepton && layer==1) {
+              lepton.SFpcal=energy/vec_lepton.P();
+              lepton.m2pcal=(m2u+m2v+m2w)/3;
+            }
+
+          if(pindex==plepton && layer==4) {
+              lepton.SFecin=energy/vec_lepton.P();
+              lepton.m2ecin=(m2u+m2v+m2w)/3;
+            }
+          if(pindex==plepton && layer==7) {
+              lepton.SFecout=energy/vec_lepton.P();
+              lepton.m2ecout=(m2u+m2v+m2w)/3;
+            }
+
+      }
+
+      
+      m_log->Debug("Variables obtained from calorimeter bank");
+
+      return lepton;
+
+  }
+
+  double LeptonIDFilter::CalculateScore(LeptonIDVars lepton_vars) const{
+
+      //Get TMVA reader
+      TMVA::Reader *readerTMVA = new TMVA::Reader( "!Color:!Silent" );
```

Comment:
Having `!Silent` is quite verbose, and should only happen when the log level is low enough. Perhaps use `m_log->GetLevel()` to control this:
```suggestion
      TString silent_opt = m_log->GetLevel() <= Logger::Level::trace ? "!Silent" : "Silent";
      TMVA::Reader *readerTMVA = new TMVA::Reader( "!Color" + silent_opt );
```